File: lua/test/pid_neuro.lua
#!/usr/bin/env lua

package.path = arg[0]:sub(0, -arg[0]:match("([^/\\]*)$"):len() - 1) .. "?.lua;" .. package.path
local test = require("test")
local a = require("liba")
test.dir(getmetatable(a.pid_neuro))
local ctx = a.pid_neuro.new()
assert(ctx)
test.dir(getmetatable(ctx))
assert(type(a.pid_neuro.kpid(ctx, 4, 3, 2, 1)) == "userdata")
assert(type(a.pid_neuro.wpid(ctx, 3, 2, 1)) == "userdata")
assert(type(a.pid_neuro.run(ctx, 1, 0)) == "number")
assert(type(a.pid_neuro.inc(ctx, 1, 0)) == "number")
assert(type(a.pid_neuro.zero(ctx)) == "userdata")
ctx:kpid(4, 3, 2, 1):wpid(3, 2, 1):zero()
assert(type(ctx:run(1, 0)) == "number")
assert(type(ctx:inc(1, 0)) == "number")
assert(type(ctx.k) == "number")
assert(type(ctx.wp) == "number")
assert(type(ctx.wi) == "number")
assert(type(ctx.wd) == "number")
assert(type(ctx.kp) == "number")
assert(type(ctx.ki) == "number")
assert(type(ctx.kd) == "number")
assert(type(ctx.outmax) == "number")
assert(type(ctx.outmin) == "number")
assert(type(ctx.out) == "number")
assert(type(ctx.fdb) == "number")
assert(type(ctx.err) == "number")
assert(type(ctx.ec) == "number")
ctx.outmin = -10
ctx.outmax = 10
ctx.kp = 3
ctx.ki = 2
ctx.kd = 1
ctx.wp = 3
ctx.wi = 2
ctx.wd = 1
ctx.k = 0
ctx.__name = nil
assert(ctx.__name)
ctx.__index = nil
assert(ctx.__index)
ctx.__newindex = nil
assert(ctx.__newindex)
---@class a.pid_neuro
---@field __name string
---@field __index table
---@field __newindex table

File: javascript/test/pid_neuro.js
#!/usr/bin/env node
const liba = require("../liba");
function assert(expr) {
    if (!eval(expr)) {
        throw Error(expr);
    }
}
function do_leak_check() {
    for (const arg of arguments) {
        arg.delete();
    }
    if (liba.do_leak_check) {
        liba.do_leak_check();
    }
}
var ctx = new liba.pid_neuro();
ctx.wpid(1, 0, 0).kpid(10, 1, 0.1, 1);
console.log(ctx.zero().run(10, 0));
console.log(ctx.zero().inc(10, 0));
assert("ctx.k != undefined");
assert("ctx.kp != undefined");
assert("ctx.ki != undefined");
assert("ctx.kd != undefined");
assert("ctx.wp != undefined");
assert("ctx.wi != undefined");
assert("ctx.wd != undefined");
assert("ctx.outmax != undefined");
assert("ctx.outmin != undefined");
assert("ctx.out != undefined");
assert("ctx.fdb != undefined");
assert("ctx.err != undefined");
assert("ctx.ec != undefined");
ctx.k = 0;
ctx.kp = 0;
ctx.ki = 0;
ctx.kd = 0;
ctx.wp = 0;
ctx.wi = 0;
ctx.wd = 0;
ctx.outmax = 0;
ctx.outmin = 0;
do_leak_check(ctx);
